File: bindings/go/backend/feature_flag.go
package backend

import "context"

type FeatureFlagType string

const (
	FeatureFlagBool    FeatureFlagType = "bool"
	FeatureFlagString  FeatureFlagType = "string"
	FeatureFlagInteger FeatureFlagType = "int"
	FeatureFlagDecimal FeatureFlagType = "decimal"
)

type FeatureFlagScopeKind string

const (
	FeatureFlagScopeSystem FeatureFlagScopeKind = "system"
	FeatureFlagScopeTenant FeatureFlagScopeKind = "tenant"
	FeatureFlagScopeUser   FeatureFlagScopeKind = "user"
	FeatureFlagScopeEntity FeatureFlagScopeKind = "entity"
)

type FeatureFlagValue struct {
	Type         FeatureFlagType
	BoolValue    bool
	StringValue  string
	IntValue     int64
	DecimalValue float64
}

func BoolFeatureFlagValue(value bool) FeatureFlagValue {
	return FeatureFlagValue{Type: FeatureFlagBool, BoolValue: value}
}

func StringFeatureFlagValue(value string) FeatureFlagValue {
	return FeatureFlagValue{Type: FeatureFlagString, StringValue: value}
}

func IntegerFeatureFlagValue(value int64) FeatureFlagValue {
	return FeatureFlagValue{Type: FeatureFlagInteger, IntValue: value}
}

func DecimalFeatureFlagValue(value float64) FeatureFlagValue {
	return FeatureFlagValue{Type: FeatureFlagDecimal, DecimalValue: value}
}

func (v FeatureFlagValue) AsBool() (bool, bool) {
	if v.Type != FeatureFlagBool {
		return false, false
	}
	return v.BoolValue, true
}

func (v FeatureFlagValue) AsString() (string, bool) {
	if v.Type != FeatureFlagString {
		return "", false
	}
	return v.StringValue, true
}

func (v FeatureFlagValue) AsInteger() (int64, bool) {
	if v.Type != FeatureFlagInteger {
		return 0, false
	}
	return v.IntValue, true
}

func (v FeatureFlagValue) AsDecimal() (float64, bool) {
	if v.Type != FeatureFlagDecimal {
		return 0, false
	}
	return v.DecimalValue, true
}

type FeatureFlagDefinition struct {
	Name         string
	Type         FeatureFlagType
	DefaultValue FeatureFlagValue
	Scope        FeatureFlagScopeKind
	Owner        *string
	Description  *string
	Expires      *string
}

type FeatureFlagRegisterDefinitionResult struct {
	RegisteredNew bool
	Definition    FeatureFlagDefinition
}

type FeatureFlagEvaluationContext struct {
	TenantID   *string
	UserID     *string
	EntityType *string
	EntityID   *string
}

type FeatureFlagEvaluationRequest struct {
	Name    string
	Context FeatureFlagEvaluationContext
}

type FeatureFlagStore interface {
	RegisterDefinition(ctx context.Context, backend Backend, definition FeatureFlagDefinition) (FeatureFlagRegisterDefinitionResult, error)

	RegisterDefinitionTx(ctx context.Context, tx Transaction, definition FeatureFlagDefinition) (FeatureFlagRegisterDefinitionResult, error)

	InspectDefinition(ctx context.Context, backend Backend, name string) (*FeatureFlagDefinition, error)

	InspectDefinitionTx(ctx context.Context, tx Transaction, name string) (*FeatureFlagDefinition, error)

	Evaluate(ctx context.Context, backend Backend, request FeatureFlagEvaluationRequest) (FeatureFlagValue, error)

	EvaluateTx(ctx context.Context, tx Transaction, request FeatureFlagEvaluationRequest) (FeatureFlagValue, error)
}

File: bindings/rust/feature_flag.rs
use crate::backend::{Backend, BackendResult};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FeatureFlagType {
    Bool,
    String,
    Integer,
    Decimal,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FeatureFlagScopeKind {
    System,
    Tenant,
    User,
    Entity,
}

#[derive(Debug, Clone, PartialEq)]
pub enum FeatureFlagValue {
    Bool(bool),
    String(String),
    Integer(i64),
    Decimal(f64),
}

impl FeatureFlagValue {
    pub fn flag_type(&self) -> FeatureFlagType {
        match self {
            FeatureFlagValue::Bool(_) => FeatureFlagType::Bool,
            FeatureFlagValue::String(_) => FeatureFlagType::String,
            FeatureFlagValue::Integer(_) => FeatureFlagType::Integer,
            FeatureFlagValue::Decimal(_) => FeatureFlagType::Decimal,
        }
    }

    pub fn as_bool(&self) -> Option<bool> {
        match self {
            FeatureFlagValue::Bool(value) => Some(*value),
            _ => None,
        }
    }

    pub fn as_string(&self) -> Option<&str> {
        match self {
            FeatureFlagValue::String(value) => Some(value),
            _ => None,
        }
    }

    pub fn as_integer(&self) -> Option<i64> {
        match self {
            FeatureFlagValue::Integer(value) => Some(*value),
            _ => None,
        }
    }

    pub fn as_decimal(&self) -> Option<f64> {
        match self {
            FeatureFlagValue::Decimal(value) => Some(*value),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct FeatureFlagDefinition {
    pub name: String,
    pub flag_type: FeatureFlagType,
    pub default_value: FeatureFlagValue,
    pub scope: FeatureFlagScopeKind,
    pub owner: Option<String>,
    pub description: Option<String>,
    pub expires: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct FeatureFlagRegisterDefinitionResult {
    pub registered_new: bool,
    pub definition: FeatureFlagDefinition,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct FeatureFlagEvaluationContext {
    pub tenant_id: Option<String>,
    pub user_id: Option<String>,
    pub entity_type: Option<String>,
    pub entity_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FeatureFlagEvaluationRequest {
    pub name: String,
    pub context: FeatureFlagEvaluationContext,
}

pub trait FeatureFlagStore<B: Backend> {
    fn register_definition(
        &self,
        backend: &B,
        definition: &FeatureFlagDefinition,
    ) -> BackendResult<FeatureFlagRegisterDefinitionResult>;

    fn register_definition_tx(
        &self,
        tx: &mut B::Tx,
        definition: &FeatureFlagDefinition,
    ) -> BackendResult<FeatureFlagRegisterDefinitionResult>;

    fn inspect_definition(
        &self,
        backend: &B,
        name: &str,
    ) -> BackendResult<Option<FeatureFlagDefinition>>;

    fn inspect_definition_tx(
        &self,
        tx: &mut B::Tx,
        name: &str,
    ) -> BackendResult<Option<FeatureFlagDefinition>>;

    fn evaluate(
        &self,
        backend: &B,
        request: &FeatureFlagEvaluationRequest,
    ) -> BackendResult<FeatureFlagValue>;

    fn evaluate_tx(
        &self,
        tx: &mut B::Tx,
        request: &FeatureFlagEvaluationRequest,
    ) -> BackendResult<FeatureFlagValue>;
}
